avocado/core/nrunner.py

```diff
@@ -391,8 +391,14 @@ def run(self):
 
         yield self.prepare_status('started')
         most_current_execution_state_time = None
-        while process.poll() is None:
-            time.sleep(RUNNER_RUN_CHECK_INTERVAL)
+        timeout = RUNNER_RUN_CHECK_INTERVAL
+        while process.returncode is None:
+            time.sleep(timeout)
+            try:
+                stdout, stderr = process.communicate(timeout=timeout)
```

Comment:
I was trying to understand what kind of magic happens here, given that `PIPE`s are still being used, and they have an inherent size limitation. Turns out the `Popen._communicate` method does something similar to `avocado.utils.process.run`, that is, it starts a thread to drain data from the pipe and put it into a larger buffer.

When running `avocado-runner-exec-test` with this patch, it seems to be able to keep the PIPEs below capacity and the complete data is kept.  But something weird still seems to be happening, as I could not get an `avocado run --test-runner=nrunner ./tests/vmx` to complete successfully.

Reply:
Hi @clebergnu, the thing is: we have two different problems: One is related to subprocess reading a large amount of data from stdout. For this one, I think this "magic" (process.communicate()) is solving this problem well.

The second one is about sending a large amount of data through asyncio streams. I did some pushes here, with a few more commits. My proposal is here, although I think we should improve this in the future.

Reply:
Right, I was missing that second part of the problem.  Thanks for identifying it.  Now, I think you proposed solution is a valid one.

But, I would like to *try* to avoid changing the "wire protocol" and adding a special case.  What I mean is, the content that is output to STDOUT when running `avocado-runner runnable-run ...` is the same data gets sent over a socket.  Some nrunner use cases, such as _Disconnected test execution, so that results can be saved to a device and collected by the runner_ (from https://avocado-framework.readthedocs.io/en/83.0/guides/contributor/chapters/runners.html#motivation).  One example of such a use case is a task gets run on a *disconnected* virtual machine and has its stdout sent to a file (say in a virtio-fs based filesystem).  The same content of the file could be given to a network status server, with no extra size headers (`cat /path/to/virtiofs/status_messages | nc statuserver:port`).

I'll time-box an experiment (not more than 1 day), and try to add reliability to the reading of the status messages without the change in protocol.

Reply:
I got the motivation and it is a valid one.

But IIUC, today we don't support "disconnected test execution", or do we support? I could not find any code related to this, and that is why I was focused only on solving the current issue. So, my understanding is that this is a different issue/feature request. Or I'm missing something?

The change in the wire protocol is minimal, and IMO sending the size upfront is better than reading the full buffer, wait for a `\n`. I was thinking that if the 'stdout' had `\n` could give us some trouble. But I didn't stress enough.

Anyway, I'm curious to see how we are going to solve this without reading the big buffer in chunks since asyncio Streams has some limitations.